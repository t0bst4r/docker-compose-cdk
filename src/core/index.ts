export * from './app';
export * from './project';
export * from './resource';
export * from './stage';
export * from './synth-metadata';
