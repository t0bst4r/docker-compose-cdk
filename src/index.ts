export * from './core';
export * from './naming';
export * from './resources';
