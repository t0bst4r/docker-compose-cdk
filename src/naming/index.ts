export * from './default-naming-strategy';
export * from './naming-strategy';
