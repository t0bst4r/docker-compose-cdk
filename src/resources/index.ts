export * from './config';
export * from './network';
export * from './service';
export * from './service-props';
export * from './volume';
