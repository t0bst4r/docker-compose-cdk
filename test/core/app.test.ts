import * as fs from 'fs';
import { tmpdir } from 'os';
import * as path from 'path';
import { Construct } from 'constructs';
import { App, DockerImage, Network, PortMapping, Project, RestartPolicy, Service, Volume } from '../../src';

describe('App', () => {


  it('should be able to synthesize two simple stacks in memory', () => {
    const app = new App();

    const firstProject = new Project(app, 'First');
    new Nginx(firstProject, 'Nginx');

    const secondProject = new Project(app, 'Second');
    new Wordpress(secondProject, 'Wordpress');
    expect(app.synth()).toMatchSnapshot();
  });

  it('should be able to synthesize two simple stacks to the filesystem', () => {
    const app = new App();

    const firstProject = new Project(app, 'First');
    new Nginx(firstProject, 'Nginx');

    const secondProject = new Project(app, 'Second');
    new Wordpress(secondProject, 'Wordpress');

    const outdir = fs.mkdtempSync(path.join(tmpdir(), 'docker-compose-cdk-test-app-synth-'));
    const projects = app.synthToDir(outdir);
    expect(projects).toEqual({
      projects: [
        {
          composeHash: 'fecfd64745042024',
          name: 'first',
        },
        {
          composeHash: 'a9266a4abc31b34c',
          name: 'second',
        },
      ],
    });

    const fileList = fs.readdirSync(outdir, { withFileTypes: true, recursive: true })
      .filter(file => file.isFile())
      .map(file => `${file.parentPath}/${file.name}`);

    expect(fileList.map(filePath => path.relative(outdir, filePath))).toEqual([
      'first.fecfd64745042024/docker-compose.json',
      'second.a9266a4abc31b34c/docker-compose.json',
    ]);

    const fileContents = fileList.map(filePath => fs.readFileSync(filePath, { encoding: 'utf-8' }));
    expect(fileContents).toMatchSnapshot();

    fs.rmSync(outdir, { recursive: true, force: true });
  });
});


class Nginx extends Construct {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    const network = new Network(this, 'Network');
    const volume = new Volume(this, 'HtmlData');
    new Service(this, 'Default', {
      image: DockerImage.hub('nginx', 'alpine'),
      restart: RestartPolicy.UNLESS_STOPPED,
      volumes: [
        { source: volume, target: '/usr/share/nginx/html' },
      ],
      ports: [
        PortMapping.http(),
        PortMapping.https(),
      ],
      networks: [{ network }],
    });
  }
}

class Wordpress extends Construct {

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const databaseName = 'exampledb';
    const databaseUser = 'exampleuser';
    const databasePassword = 'examplepass';

    const network = new Network(this, 'Network');

    const dbData = new Volume(this, 'DatabaseData');
    const db = new Service(this, 'DatabaseService', {
      image: DockerImage.hub('mysql', '8.0'),
      restart: RestartPolicy.UNLESS_STOPPED,
      environment: {
        MYSQL_DATABASE: databaseName,
        MYSQL_USER: databaseUser,
        MYSQL_PASSWORD: databasePassword,
        MYSQL_RANDOM_ROOT_PASSWORD: '1',
      },
      volumes: [
        { source: dbData, target: '/var/lib/mysql' },
      ],
      networks: [{ network }],
    });

    const wordpressData = new Volume(this, 'WordpressData');
    new Service(this, 'Default', {
      image: DockerImage.hub('wordpress'),
      restart: RestartPolicy.UNLESS_STOPPED,
      environment: {
        WORDPRESS_DB_HOST: db.dnsNames[0],
        WORDPRESS_DB_NAME: databaseName,
        WORDPRESS_DB_USER: databaseUser,
        WORDPRESS_DB_PASSWORD: databasePassword,
      },
      volumes: [
        { source: wordpressData, target: '/var/www/html' },
      ],
      networks: [{ network }],
    });
  }

}
