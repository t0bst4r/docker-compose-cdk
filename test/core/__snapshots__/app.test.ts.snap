// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should be able to synthesize two simple stacks in memory 1`] = `
{
  "projects": [
    {
      "compose": {
        "name": "first",
        "networks": {
          "nginx_network": {
            "driver": "bridge",
            "driver_opts": undefined,
            "ipam": undefined,
          },
        },
        "services": {
          "nginx": {
            "command": undefined,
            "configs": undefined,
            "deploy": {
              "mode": "replicated",
              "replicas": 1,
            },
            "devices": undefined,
            "environment": undefined,
            "image": "nginx:alpine",
            "networks": {
              "nginx_network": {
                "ipv4_address": undefined,
              },
            },
            "ports": [
              "80:80",
              "443:443",
            ],
            "privileged": false,
            "profiles": undefined,
            "restart": "unless-stopped",
            "user": undefined,
            "volumes": [
              {
                "read_only": undefined,
                "source": "nginx_htmldata",
                "target": "/usr/share/nginx/html",
                "type": "volume",
              },
            ],
          },
        },
        "volumes": {
          "nginx_htmldata": {},
        },
      },
      "composeHash": "fecfd64745042024",
      "name": "first",
    },
    {
      "compose": {
        "name": "second",
        "networks": {
          "wordpress_network": {
            "driver": "bridge",
            "driver_opts": undefined,
            "ipam": undefined,
          },
        },
        "services": {
          "wordpress": {
            "command": undefined,
            "configs": undefined,
            "deploy": {
              "mode": "replicated",
              "replicas": 1,
            },
            "devices": undefined,
            "environment": {
              "WORDPRESS_DB_HOST": "second-wordpress-databaseservice-1",
              "WORDPRESS_DB_NAME": "exampledb",
              "WORDPRESS_DB_PASSWORD": "examplepass",
              "WORDPRESS_DB_USER": "exampleuser",
            },
            "image": "wordpress:latest",
            "networks": {
              "wordpress_network": {
                "ipv4_address": undefined,
              },
            },
            "ports": undefined,
            "privileged": false,
            "profiles": undefined,
            "restart": "unless-stopped",
            "user": undefined,
            "volumes": [
              {
                "read_only": undefined,
                "source": "wordpress_wordpressdata",
                "target": "/var/www/html",
                "type": "volume",
              },
            ],
          },
          "wordpress_databaseservice": {
            "command": undefined,
            "configs": undefined,
            "deploy": {
              "mode": "replicated",
              "replicas": 1,
            },
            "devices": undefined,
            "environment": {
              "MYSQL_DATABASE": "exampledb",
              "MYSQL_PASSWORD": "examplepass",
              "MYSQL_RANDOM_ROOT_PASSWORD": "1",
              "MYSQL_USER": "exampleuser",
            },
            "image": "mysql:8.0",
            "networks": {
              "wordpress_network": {
                "ipv4_address": undefined,
              },
            },
            "ports": undefined,
            "privileged": false,
            "profiles": undefined,
            "restart": "unless-stopped",
            "user": undefined,
            "volumes": [
              {
                "read_only": undefined,
                "source": "wordpress_databasedata",
                "target": "/var/lib/mysql",
                "type": "volume",
              },
            ],
          },
        },
        "volumes": {
          "wordpress_databasedata": {},
          "wordpress_wordpressdata": {},
        },
      },
      "composeHash": "a9266a4abc31b34c",
      "name": "second",
    },
  ],
}
`;

exports[`App should be able to synthesize two simple stacks to the filesystem 1`] = `
[
  "{
  "networks": {
    "nginx_network": {
      "driver": "bridge"
    }
  },
  "volumes": {
    "nginx_htmldata": {}
  },
  "services": {
    "nginx": {
      "image": "nginx:alpine",
      "privileged": false,
      "restart": "unless-stopped",
      "deploy": {
        "mode": "replicated",
        "replicas": 1
      },
      "ports": [
        "80:80",
        "443:443"
      ],
      "volumes": [
        {
          "type": "volume",
          "source": "nginx_htmldata",
          "target": "/usr/share/nginx/html"
        }
      ],
      "networks": {
        "nginx_network": {}
      }
    }
  },
  "name": "first"
}",
  "{
  "networks": {
    "wordpress_network": {
      "driver": "bridge"
    }
  },
  "volumes": {
    "wordpress_databasedata": {},
    "wordpress_wordpressdata": {}
  },
  "services": {
    "wordpress_databaseservice": {
      "image": "mysql:8.0",
      "privileged": false,
      "restart": "unless-stopped",
      "deploy": {
        "mode": "replicated",
        "replicas": 1
      },
      "environment": {
        "MYSQL_DATABASE": "exampledb",
        "MYSQL_USER": "exampleuser",
        "MYSQL_PASSWORD": "examplepass",
        "MYSQL_RANDOM_ROOT_PASSWORD": "1"
      },
      "volumes": [
        {
          "type": "volume",
          "source": "wordpress_databasedata",
          "target": "/var/lib/mysql"
        }
      ],
      "networks": {
        "wordpress_network": {}
      }
    },
    "wordpress": {
      "image": "wordpress:latest",
      "privileged": false,
      "restart": "unless-stopped",
      "deploy": {
        "mode": "replicated",
        "replicas": 1
      },
      "environment": {
        "WORDPRESS_DB_HOST": "second-wordpress-databaseservice-1",
        "WORDPRESS_DB_NAME": "exampledb",
        "WORDPRESS_DB_USER": "exampleuser",
        "WORDPRESS_DB_PASSWORD": "examplepass"
      },
      "volumes": [
        {
          "type": "volume",
          "source": "wordpress_wordpressdata",
          "target": "/var/www/html"
        }
      ],
      "networks": {
        "wordpress_network": {}
      }
    }
  },
  "name": "second"
}",
]
`;
